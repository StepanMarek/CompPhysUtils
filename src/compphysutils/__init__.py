__all__ = ["crystalgen", "graphics"]
