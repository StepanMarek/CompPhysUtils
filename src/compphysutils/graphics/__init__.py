from .atom_plot import atom_colors, atom_sizes

__all__ = ["parser"]
