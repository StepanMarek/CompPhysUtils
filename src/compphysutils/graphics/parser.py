from . import hlgParser
from .post_process import postProcessCommands
import configparser

lineParseFunctions = {
    "hlg" : hlgParser.hlgLine
}

parserKwargsDefaults = {
    "hlg" : {
        "outputUnit" : "eV"
    }
}

initObjectsFunctions = {
    "hlg" : hlgParser.initParserObjects 
}

def parseFile(filename, filetype, parserKwargs=False):
    if not parserKwargs:
        parserKwargs = parserKwargsDefaults[filetype]
    file = open(filename, "r")
    datagroups = []
    parserObjects = initObjectsFunctions[filetype]()
    currentParser = lineParseFunctions[filetype]
    for line in file:
        # Read line by line
        # Can return bool False if line is to be skipped
        readGroups = currentParser(line, *parserObjects, **parserKwargs)
        if readGroups:
            for i in range(len(readGroups)):
                if len(datagroups) > i:
                    datagroups[i].append(readGroups[i])
                else:
                    datagroups.append([readGroups[i]])
    file.close()
    return datagroups

def postProcess(datagroups, command, args):
   return postProcessCommands[command](datagroups, *args)

def parseDatasetConfig(configFilename):
    cfg = configparser.ConfigParser()
    cfg.read(configFilename)
    datasets = {}
    for groupName in cfg.sections():
        if "dataset" in groupName:
            datasetName = groupName.split(".")[1]
            if "file" in cfg[groupName]:
                # Create datasets from file
                parserKwargs = cfg.get(groupName, "parser-kwargs", fallback=False)
                datasets[datasetName] = parseFile(cfg[groupName]["file"], cfg[groupName]["filetype"], parserKwargs=parserKwargs)
            elif "list" in cfg[groupName]:
                # Create dataset from list, defaultly convert to float
                # TODO : Should there be som interface to different convertors?
                datasets[datasetName] = list(map(float, cfg[groupName]["list"].split()))
            if cfg[groupName]["post-processing"]:
                commandSplit = cfg[groupName]["post-processing"].split()
                if len(commandSplit) > 1:
                    datasets[datasetName] = postProcess(datasets[datasetName], commandSplit[0], commandSplit[1:])
                else:
                    datasets[datasetName] = postProcess(datasets[datasetName], commandSplit[0], [])
    return datasets
