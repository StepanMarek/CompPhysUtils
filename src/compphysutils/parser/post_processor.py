import os
import importlib

## Search for default post-processing commands
roots = []
modFilenames = []
root, _, filenames = next(os.walk(os.path.dirname(__file__)+"/post_process"))
roots.append(root)
modFilenames.append(filenames)
## Search for custom post_process commands
if os.path.isdir(os.path.expanduser("~/.config/compphysutils/post_process")):
    root, _, filenames = next(os.walk(os.path.expanduser("~/.config/compphysutils/post_process")))
    roots.append(root)
    modFilenames.append(filenames)
## Import all commands
postProcessCommands = {}
for i in range(len(roots)):
    for filename in modFilenames[i]:
        commandName = filename.split(".")[0]
        if commandName[0:2] == "__":
            # Skip __init__.py and similar commands
            continue
        spec = importlib.util.spec_from_file_location("compphysutils.parser.post_process."+commandName, roots[i]+"/"+filename)
        mod = importlib.util.module_from_spec(spec)
        spec.loader.exec_module(mod)
        postProcessCommands[commandName] = mod.command
