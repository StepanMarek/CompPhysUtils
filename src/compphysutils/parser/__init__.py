from .parser import parseDatasetConfig, save, writeFile, readFile
