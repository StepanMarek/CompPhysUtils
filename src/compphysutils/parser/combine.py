import importlib
import os
from .parser import parseDatasetConfig
from .parser import writeParseFunctions
from .parser import writeHeaderFunctions
from .parser import writeFooterFunctions
from .savepoint import parse as savepointParse

## Search for default combine commands
roots = []
modFilenames = []
root, _, filenames = next(os.walk(os.path.dirname(__file__)+"/combine_commands"))
roots.append(root)
modFilenames.append(filenames)
## Search for custom combine commands
if os.path.isdir(os.path.expanduser("~/.config/compphysutils/combine_commands")):
    root, _, filenames = next(os.walk(os.path.expanduser("~/.config/compphysutils/combine_commands")))
    roots.append(root)
    modFilenames.append(filenames)
## Import all commands
commands = {}
for i in range(len(roots)):
    for filename in modFilenames[i]:
        commandName = filename.split(".")[0]
        if commandName[0:2] == "__":
            # Skip __init__.py and similar commands
            continue
        spec = importlib.util.spec_from_file_location("compphysutils.parser.combine_commands."+commandName, roots[i]+"/"+filename)
        mod = importlib.util.module_from_spec(spec)
        spec.loader.exec_module(mod)
        commands[commandName] = mod.command

def runGroupData(cfg, datasets, cfgFileName):
    if "data" in cfg:
        # First, load data from datasetfiles
        datasetfiles = cfg["data"].get("datasetfiles", False)
        if datasetfiles:
            for datasetFileName in datasetfiles.split():
                datasets.update(parseDatasetConfig(datasetFileName))
    # If there are in place defined datasets, they take priority
    datasets.update(parseDatasetConfig(cfgFileName))
    # If there is no data group, this is all that is done
    if not "data" in cfg:
        return datasets
    # Now, progress to the combine command
    if "combine" in cfg["data"]:
        combineCommands = cfg.get("data", "combine").split("\n")
        for commandLine in combineCommands:
            commandSplitLine = commandLine.split()
            commandName = commandSplitLine[0]
            datasets = commands[commandName](datasets, commandSplitLine[1:])
    # Finally, handle savepoint in the combine context
    if "savepoint" in cfg["data"]:
        savepointParse(cfg["data"].get("savepoint"), "combine", datasets, writeParseFunctions, writeHeaderFunctions, writeFooterFunctions, "data_combine.out")
    # End by returning the (changed) datasets
    return datasets
