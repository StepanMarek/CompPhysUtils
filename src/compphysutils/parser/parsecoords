#!/usr/bin/python
import argparse
import sys
from compphysutils.parser.parser import initObjectsFunctions, lineParseFunctions, parserArgsDefaults, writeParseFunctions, writeHeaderFunctions, writeFooterFunctions

ap = argparse.ArgumentParser(description="Parses coordinate files from one format to another.")
ap.add_argument("source_filename", help="Filename of the source file that contains the coordinates in the source_format.")
ap.add_argument("source_format", help="The format of the source_filename file.")
ap.add_argument("target_format", help="Format in which the coordinates are outputed. The output is written into stdout.")
ap.add_argument("--source_parser_args", nargs="*", default=False, help="Optional args to be given to the source parser.")
ap.add_argument("--target_parser_args", nargs="*", default=False, help="Optional args to be given to the target parser.")

args = ap.parse_args()
# Get parser arguments
if args.source_parser_args:
    sourceArgs = args.source_parser_args
else:
    sourceArgs = parserArgsDefaults[args.source_format]
if args.target_parser_args:
    targetArgs = args.target_parser_args
else:
    targetArgs = parserArgsDefaults[args.target_format]
# Initiate parser objects
sourceObjs = initObjectsFunctions[args.source_format](sourceArgs)
targetObjs = initObjectsFunctions[args.target_format](targetArgs)
# Read source file
sourceFile = open(args.source_filename, "r")
dataset = []
for line in sourceFile:
    datarow = lineParseFunctions[args.source_format](line, *sourceObjs)
    if datarow:
        for i in range(len(datarow)):
            if len(dataset) > i:
                dataset[i].append(datarow[i])
            else:
                dataset.append([datarow[i]])
sourceFile.close()
# Write target file to stdout
if writeHeaderFunctions[args.target_format]:
    print(writeHeaderFunctions[args.target_format](dataset))
for i in range(len(dataset[0])):
    datarow = []
    for j in range(len(dataset)):
        datarow.append(dataset[j][i])
    print(writeParseFunctions[args.target_format](datarow))
if writeFooterFunctions[args.target_format]:
    print(writeFooterFunctions[args.target_format](dataset))

sys.exit(0)
