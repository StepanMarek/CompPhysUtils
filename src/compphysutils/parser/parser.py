from .post_processor import postProcessCommands
from .savepoint import datasetParse as savepointDatasetParse
import configparser
import os
import importlib
import re

parserModules = {}
# Check for custom parser modules
customParsers = []
if os.path.isdir(os.path.expanduser("~/.config/compphysutils/parsers")):
    root, _, filenames = next(os.walk(os.path.expanduser("~/.config/compphysutils/parsers")))
    customParsers = list(map(lambda f: root+"/"+f, filenames))
# Check for default parser modules
root, _, filenames = next(os.walk(os.path.dirname(__file__)+"/parsers"))
defaultParsers = list(map(lambda f: root+"/"+f, filenames))
# Ideally, custom parsers should overwrite default parsers, not sure if this works
rexp = re.compile(".*/(.*)\.py")
for parserFileName in (defaultParsers + customParsers):
    parserType = rexp.search(parserFileName).group(1)
    if parserType[0:2] == "__":
        # Not a parser, but just a init or other module
        continue
    spec = importlib.util.spec_from_file_location("compphysutils.parser.parsers."+parserType, parserFileName)
    mod = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(mod)
    parserModules[parserType] = mod
lineParseFunctions = {}
writeParseFunctions = {}
writeHeaderFunctions = {}
writeFooterFunctions = {}
parserArgsDefaults = {}
initObjectsFunctions = {}
for parserName in parserModules:
    lineParseFunctions[parserName] = parserModules[parserName].line
    parserArgsDefaults[parserName] = parserModules[parserName].argDefaults
    initObjectsFunctions[parserName] = parserModules[parserName].initParserObjects
    if hasattr(parserModules[parserName], "writeLine"):
        writeParseFunctions[parserName] = parserModules[parserName].writeLine
    else:
        writeParseFunctions[parserName] = False
    if hasattr(parserModules[parserName], "writeHeaders"):
        writeHeaderFunctions[parserName] = parserModules[parserName].writeHeaders
    else:
        writeHeaderFunctions[parserName] = False
    if hasattr(parserModules[parserName], "writeFooters"):
        writeFooterFunctions[parserName] = parserModules[parserName].writeFooters
    else:
        writeFooterFunctions[parserName] = False

def parseFile(filename, filetype, parserArgs=False):
    if not parserArgs:
        parserArgs = parserArgsDefaults[filetype]
    file = open(filename, "r")
    datagroups = []
    parserObjects = initObjectsFunctions[filetype](parserArgs)
    currentParser = lineParseFunctions[filetype]
    for line in file:
        # Read line by line
        # Can return bool False if line is to be skipped
        readGroups = currentParser(line, *parserObjects)
        if readGroups:
            for i in range(len(readGroups)):
                if len(datagroups) > i:
                    datagroups[i].append(readGroups[i])
                else:
                    datagroups.append([readGroups[i]])
    file.close()
    return datagroups

def postProcess(datagroups, command, args):
   return postProcessCommands[command](datagroups, args)

def parseDatasetConfig(configFilename):
    cfg = configparser.ConfigParser()
    cfg.read(configFilename)
    datasets = {}
    for groupName in cfg.sections():
        if "dataset" in groupName:
            datasetName = groupName.split(".")[1]
            if "file" in cfg[groupName]:
                # Create datasets from file
                parserArgs = cfg[groupName].get("parser-args", False)
                datasets[datasetName] = parseFile(cfg[groupName]["file"], cfg[groupName]["filetype"], parserArgs=parserArgs)
            elif "list" in cfg[groupName]:
                # Create dataset from list, defaultly convert to float
                # TODO : Should there be som interface to different convertors?
                # A more sophisticated and decoupled list types
                splitList = cfg[groupName]["list"].split("\n")
                if "listtype" in cfg[groupName]:
                    if cfg[groupName]["listtype"] == "string":
                        datasets[datasetName] = list(map(lambda s: s.split(), splitList))
                    else:
                        datasets[datasetName] = list(map(lambda s: list(map(float, s.split())), splitList))
                else:
                    datasets[datasetName] = list(map(lambda s: list(map(float, s.split())), splitList))
            if "savepoint" in cfg[groupName]:
                savepointDatasetParse(cfg[groupName].get("savepoint"), "load", datasets[datasetName], writeParseFunctions, writeHeaderFunctions, writeFooterFunctions)
            if "post-process" in cfg[groupName]:
                commandSplit = cfg[groupName]["post-process"].split()
                if len(commandSplit) > 1:
                    datasets[datasetName] = postProcess(datasets[datasetName], commandSplit[0], commandSplit[1:])
                else:
                    datasets[datasetName] = postProcess(datasets[datasetName], commandSplit[0], [])
            if "savepoint" in cfg[groupName]:
                savepointDatasetParse(cfg[groupName].get("savepoint"), "post-process", datasets[datasetName], writeParseFunctions, writeHeaderFunctions, writeFooterFunctions, defFilename="data_post.out")
    return datasets
