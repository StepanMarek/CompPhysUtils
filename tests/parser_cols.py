import unittest
import compphysutils.parser as parser

class ParserColsTest(unittest.TestCase):

    def test_read(self):
        dataset = parser.readFile("parser_cols.out", "cols", parserArgs="0 1")
        self.assertEqual(dataset[0][0], 1.0)
        self.assertEqual(dataset[0][1], 2.0)
        self.assertEqual(dataset[1][0], 4.0)
        self.assertEqual(dataset[1][1], 9.0)

    def test_write(self):
        dataset = [[2.0, 4.0],[3.0, 7.0]]
        parser.writeFile("parser_cols.write", "cols", dataset)
        with open("parser_cols.write", "r") as file:
            self.assertEqual(file.readline().split(), ["2.0", "3.0"])
            self.assertEqual(file.readline().split(), ["4.0", "7.0"])

if __name__ == "__main__":
    unittest.main()
