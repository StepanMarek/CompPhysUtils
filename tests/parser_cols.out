1.0 4.0
2.0 9.0
